discovery/controllers/decorators/addon.go

```diff
@@ -267,6 +278,50 @@ func (c *Component) Reference() (ref *discoveryv1alpha1.RichReference, err error
 	return
 }
 
+func (c *Component) ExtractMetadata(ref *discoveryv1alpha1.RichReference) (metas []discoveryv1alpha1.Metadata, err error) {
+	m, err := meta.Accessor(c)
+	if err != nil {
+		return
+	}
+	metas = make([]discoveryv1alpha1.Metadata, 0)
+	annotations := m.GetAnnotations()
+	if len(annotations) == 0 {
+		return
+	}
+
+	marshalled, err := c.Unstructured.MarshalJSON()
+	if err != nil {
+		return
+	}
+	for k, v := range annotations {
+		if !strings.HasPrefix(k, ComponentLabelKeyPrefix) {
+			continue
+		}
+
+		// todo: don't convert to string, use gjson byte interface
+		objs, oerr := builder.BuildObject(strings.TrimPrefix(k, ComponentLabelKeyPrefix), v, string(marshalled))
```

Comment:
Currently, a component can belong to multiple operators. The way this is currently implemented, it looks like all operators will receive the metadata values for all component labels; e.g.

```
<prefix>/op-a: "{'goose': 'honk'}"
<prefix>/op-b: "{'gander': 'honk'}"
```

The above labels on a component will result in **both** operators -- op-a and op-b -- getting **both** metadata selections ({'goose': 'honk', 'gander': 'honk'}).

I think this is a little misleading and we should either:
- add metadata values to an operator only when the label key suffix matches the operator
- add a separate metadata key, disjoint from a specific operator's name, which describes metadata for all operators it is a component of